lib/spack/spack/patch.py

```diff
@@ -237,7 +239,8 @@ def stage(self):
 
         # The same package can have multiple patches with the same name but
         # with different contents, therefore apply a subset of the hash.
-        name = '{0}-{1}'.format(os.path.basename(self.url), fetch_digest[:7])
+        base = urllib_parse.quote(os.path.basename(self.url))
```

Comment:
Could you try doing this in `URLFetchStrategy.__init__`? I think that might address other issues where the URL is not handled properly.

Reply:
Done! `URLFetchStrategy` now has a new `file_name` member that tries to figure out an appropriate name to use. Is this what you had in mind?

Reply:
I was hoping that all you'd have to do is escape `self.url` in `URLFetchStrategy.__init__` (so you wouldn't have to retrieve any paths from the fetcher) - does that work?

Reply:
I am not sure I follow. The original problem was that the patch code tried to use the basename in a way that could fail. Now we extract it in a more robust way and fall back to the encoded full URL if that doesn't work. Do you want to always use the full URL as the filename?

Reply:
Apologies: my suggestion was incorrect. I missed that the problem was with the file that curl was saving *to*, not the URL that it was retrieving from.

I tried the following patch which worked for me:

```
diff --git a/lib/spack/spack/fetch_strategy.py b/lib/spack/spack/fetch_strategy.py
index b5ce8495e3..5604e7182e 100644
--- a/lib/spack/spack/fetch_strategy.py
+++ b/lib/spack/spack/fetch_strategy.py
@@ -336,7 +336,7 @@ def _fetch_from_url(self, url):
         partial_file = None
         if self.stage.save_filename:
             save_file = self.stage.save_filename
-            partial_file = self.stage.save_filename + '.part'
+            partial_file = urllib_parse.quote(self.stage.save_filename) + '.part'
         tty.msg('Fetching {0}'.format(url))
         if partial_file:
             save_args = ['-C',
```

I was trying extend the fix you proposed to handle other resources with this problem (i.e. not just patches). Does this seem reasonable to you? If so you can apply it here in this PR and I'll merge, or I can open a different one (your preference).

This also keeps the cosmetic name the same as before (I think the PR in its current state changes the name).

Reply:
If I add your change, I still get the same error:
```
==> Adding package autoconf@2.70 to mirror
==> Fetching file://$SPACK/mirror/autoconf/?id=05972f49ee632cd98057a3caf82ebfb9574846da-eaa3f69

==> Warning: Error while fetching autoconf@2.70
  sha256 checksum failed for $STAGE/spack-stage-0r7o7mg_/?id=05972f49ee632cd98057a3caf82ebfb9574846da
```

The problem occurs when trying to fetch the patch (because Spack/curl thinks it has already downloaded it due to the encoding issue), not when saving it.

Reply:
Got it - I was having trouble reproducing your error (now I can). I'd like to handle this in a way that resolves the issue for all possible local fetches (not just for patches). Does the following work for you:

```
diff --git a/lib/spack/spack/fetch_strategy.py b/lib/spack/spack/fetch_strategy.py
index b5ce8495e3..78cd0601ad 100644
--- a/lib/spack/spack/fetch_strategy.py
+++ b/lib/spack/spack/fetch_strategy.py
@@ -332,6 +332,12 @@ def _existing_url(self, url):
         return curl.returncode == 0
 
     def _fetch_from_url(self, url):
+        components = urllib_parse.urlsplit(url)
+        if components.scheme == 'file':
+            path = url[len('file://'):]
+            url = 'file://' + urllib_parse.quote(path)
+
+        import pdb; pdb.set_trace()
         save_file = None
         partial_file = None
         if self.stage.save_filename:
```

?

Reply:
That indeed mostly fixed it. `_existing_url` still used the unescaped URL, so I had to move it to `candidate_urls`. I also rolled back the basename handling, so file names should remain the same.